throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,146 @@
+package throttle
+
+import (
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/loomnetwork/go-loom"
+	"github.com/loomnetwork/go-loom/plugin/contractpb"
+	"github.com/loomnetwork/loomchain"
+	"github.com/loomnetwork/loomchain/auth"
+	udw "github.com/loomnetwork/loomchain/builtin/plugins/user_deployer_whitelist"
+	"github.com/loomnetwork/loomchain/vm"
+	"github.com/pkg/errors"
+)
+
+var ErrTxLimitReached = errors.New("tx limit reached, try again later")
+
+type ContractTxLimiterConfig struct {
+	// Enables the middleware
+	Enabled bool
+	// Number of seconds each refresh lasts
+	RefreshInterval int64
+}
+
+func DefaultContractTxLimiterConfig() *ContractTxLimiterConfig {
+	return &ContractTxLimiterConfig{
+		Enabled:         false,
+		RefreshInterval: 15 * 60,
+	}
+}
+
+// Clone returns a deep clone of the config.
+func (c *ContractTxLimiterConfig) Clone() *ContractTxLimiterConfig {
+	if c == nil {
+		return nil
+	}
+	clone := *c
+	return &clone
+}
+
+type contractTxLimiter struct {
+	// contract_address to limiting parametres structure
+	contractToTierMap map[string]udw.Tier
+	// track of no. of txns in previous blocks per contract
+	contractToBlockTrx map[string]map[int64]int64
+	// to refresh the contractToTierMap when configuration changes
+	lastUpdated int64
+}
+
+var TxLimiter *contractTxLimiter
+
+func (txl *contractTxLimiter) isAccountLimitReached(contractAddr loom.Address, curBlockHeight int64) bool {
```

Comment:
This is overly complicated. For each contract address keep track of the start of the block range, and a count of txs within that range, each time a tx comes in increment the counter, when the block range ends reset the counter and set the start of the block range to the current height.